ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -217,6 +250,36 @@ private void handleForkChoiceResult(
             maybeState -> calculatePayloadAttributes(blockSlot, epoch, maybeState), eventThread);
   }
 
+  private Optional<Bytes32> getExecutionPayloadHeadFromPow(UInt64 slot) {
+    MiscHelpersMerge miscHelpers = ((MiscHelpersMerge) spec.atSlot(slot).miscHelpers());
+
+    return recentChainData
+        .getBestState()
+        .flatMap(
+            state -> {
+              SpecConfigMerge specConfig =
+                  SpecConfigMerge.required(spec.getSpecConfig(spec.getCurrentEpoch(state)));
+
+              if (!miscHelpers.isMergeComplete(state)) {
+                MergeTransitionHelpers mergeTransitionHelpers =
+                    spec.atSlot(slot).getMergeTransitionHelpers().orElseThrow();
+                final boolean isTerminalBlockHashSet = !specConfig.getTerminalBlockHash().isZero();
+                final boolean isActivationEpochReached =
+                    miscHelpers
+                        .computeEpochAtSlot(slot)
+                        .isGreaterThanOrEqualTo(specConfig.getTerminalBlockHashActivationEpoch());
+                if (isTerminalBlockHashSet && !isActivationEpochReached) return Optional.empty();
+
+                Optional<PowBlock> terminalPowBlock =
+                    mergeTransitionHelpers.getTerminalPowBlock(executionEngineChannel).join();
```

Comment:
We usually try really hard to avoid using `.join()` because it can cause issues when threads are blocked, including some surprising deadlocks at times.